freeipa/src/main/resources/freeipa-salt/salt/freeipa/scripts/initdnarange.py

```diff
@@ -0,0 +1,156 @@
+from ipalib import api
+from ipaserver.install import replication
+import logging
+from random import randint
+from time import sleep
+
+
+class RangeSet:
+    def __init__(self, elements):
+        self.ranges = list(elements)
+
+    def __iter__(self):
+        return iter(self.ranges)
+
+    def __repr__(self):
+        return 'RangeSet: %r' % self.ranges
+
+    def has(self, tup):
+        for pos, i in enumerate(self.ranges):
+            if i[0] <= tup[0] and i[1] >= tup[1]:
+                return pos, i
+        raise ValueError('Invalid range or overlapping range')
+
+    def minus(self, tup):
+        pos, (x, y) = self.has(tup)
+        out = []
+        if x < tup[0]:
+            out.append((x, tup[0] - 1))
+        if y > tup[1]:
+            out.append((tup[1] + 1, y))
+        self.ranges[pos:pos + 1] = out
+
+    def __sub__(self, r):
+        r1 = RangeSet(self)
+        for i in r: r1.minus(i)
+        return r1
+
+    def sub(self, r):  # inplace subtraction
+        for i in r:
+            self.minus(i)
+
+    def subWithResult(self, result, r):  # unfortunately ipa console won't work with the __sub__ solution, couldn't initialize RangeSet
+        result.ranges = self.ranges
+        for i in r:
+            result.minus(i)
+        return result
+
+
+class DnaRangeIniter:
+    """Needed a class as FreeIPA console couldn't resolved method to method calls"""
+
+    MIN_RANGE_SIZE = 1000
```

Comment:
I fear this isn't sufficient. If 1000 users/groups are added then how does a new range get reallocated?

Reply:
it will ask for a range from another instance after exhaustion. I noticed during testing that sometimes there are ranges with 1 free ID during the default behavior. This is a mitigation to avoid assigning too small ranges in the first place.